README.md
# Ship 'til You Flip
Ludum Dare #LDJAM 32 Game Submission

## Changelog
Running log of activity available in the [changelog](CHANGELOG.md)

## Getting Started
### Installing
On the first time you set up your environment, run:

```bash
npm install
```

### Testing
For development, run:

```bash
npm test
```

### Deploying
To package code, run:

```bash
npm run deploy
```

The resulting distributables can be packaged and tagged with a new verson.

### Updating Engine
The engine integrated is [PlayCanvas](https://playcanvas.com). See their [repo](https://github.com/playcanvas/engine) for instructions on building from source. Replace `engine/playcanvas.js` with the newly built source to distribute on a later engine.

## License

This game is released under the [MIT](http://opensource.org/licenses/MIT) license. See LICENSE file.


CHANGELOG.md
### 03:30 Baseline
- completed baseline with WebPack
- integration with PlayCanvas decided

### 04:00 Name decided!
```bash
    ____  _     _         _ _   _ _  __   __            _____ _ _       
   / ___|| |__ (_)_ __   ( ) |_(_) | \ \ / /__  _   _  |  ___| (_)_ __  
   \___ \| '_ \| | '_ \  |/| __| | |  \ V / _ \| | | | | |_  | | | '_ \ 
    ___) | | | | | |_) |   | |_| | |   | | (_) | |_| | |  _| | | | |_) |
   |____/|_| |_|_| .__/     \__|_|_|   |_|\___/ \__,_| |_|   |_|_| .__/ 
                 |_|                                               |_|    
```

### 10:00 First game 3D assets are in!
![Cockpit view](https://cloud.githubusercontent.com/assets/9714275/7215343/4a35a232-e5cb-11e4-9c9f-06d547506703.png)
![Unconventional weapon](https://cloud.githubusercontent.com/assets/9714275/7215341/4a327706-e5cb-11e4-8425-0a06a3e7dbf6.png)
![Ship base model](https://cloud.githubusercontent.com/assets/9714275/7215342/4a34bb10-e5cb-11e4-9732-1b0dedd9ee44.png)

### 12:00 PlayCanvas integrated with Baseline
- import script wraps pre-built engine

### 12:30 Phase 1 complete 3D assets
![Cockpit view](https://cloud.githubusercontent.com/assets/9714275/7215766/2312e5cc-e5dd-11e4-8d3d-4c3eca6455c3.png)
![Final ship](https://cloud.githubusercontent.com/assets/9714275/7215768/2313a55c-e5dd-11e4-94bf-83daf1eaa760.png)
![Una Mesa](https://cloud.githubusercontent.com/assets/9714275/7215767/2313723a-e5dd-11e4-9527-bfc64173bfd6.png)

### 13:00 Proven workflow with basic 3D rendering
- rotating cube added to 3D space in engine

### 14:30 First wave of enemies!
![Enemy Ship #1](https://cloud.githubusercontent.com/assets/1331877/7221411/cae722b8-e6e1-11e4-93d7-b2ec0743cd15.png)
![Enemy Ship #2](https://cloud.githubusercontent.com/assets/1331877/7221412/ce77c338-e6e1-11e4-89c6-c9a033657eb4.png)

### 20:50 A third enemy model
![Enemy Ship #3](https://cloud.githubusercontent.com/assets/1331877/7221413/cfcad7d4-e6e1-11e4-83ff-0fd8fc281afa.png)

### 21:30 And a fourth enemy model
![Enemy Ship #4](https://cloud.githubusercontent.com/assets/1331877/7221414/d1456ad4-e6e1-11e4-8040-ffd1c13209f7.png)

### 22:00 Asteroids!
![Asteroids](https://cloud.githubusercontent.com/assets/1331877/7221415/d30c04d6-e6e1-11e4-8478-7c4f4174f1b2.png)

### 26:00 Successfully imported model from Blender
- asteroid asset imported into game
- player ship imported into game

### 34:00 Burrnnt ouut!
Hey ho, a lot was learned this Ludum Dare.

What an exhilarating experience, to be continued...


webpack.config.js
var path = require('path'),
  webpack = require('webpack'),
  htmlPlugin = require('html-webpack-plugin')

var srcRoot = 'src/'

module.exports = {
  cache: true,
  progress: true,
  colors: true,

  entry: [path.join(__dirname, srcRoot, 'game.js')],

  output: {
    path: __dirname,
    filename: 'bundle.js'
  },

  module: {
    loaders: [
      { test: /\.jpg$/, loader: 'file-loader' },
      { test: /\.png$/, loader: 'url-loader?mimetype=image/png' },
      { test: /\.js$/, exclude: /node_modules/, loader: 'babel-loader' }
    ]
  },

  resolve: {
    alias: { 'engine': path.join(__dirname, 'engine/import.js') }
  },

  plugins: [
    new webpack.NoErrorsPlugin(),
    new webpack.optimize.CommonsChunkPlugin(['node_modules', 'engine'], 'engine.bundle.js'),
    new htmlPlugin({ title: "Ship 'til You Flip", template: path.join(srcRoot, 'index.html') })
  ]
}


src/models/entities.js
import engine from 'engine'

// assets paths config
export const assets = {
  'shwip': 'assets/player/shwip/shwip.json',
  'asteroid_1': 'assets/asteroids/asteroid_1/asteroid_1.json'
}

/**
 *  Default camera entity
 */
export function Camera () {
  var camera = new engine.Entity()

  camera.addComponent('camera', {
    clearColor: new engine.Color(0.1, 0.1, 0.1)
  })

  return camera
}

/**
 *  Default lighting entity
 */
export function Light () {
  var light = new engine.Entity()
  light.addComponent("light", {
    type: "spot",
    color: new engine.Color(1, 1, 1),
    outerConeAngle: 60,
    innerConeAngle: 40,
    range: 100,
    intensity: 1,
    castShadows: true,
    shadowBias: 0.005,
    shadowResolution: 2048
  })

  return light
}

/**
 *  Player entity
 *  @params resource
 */
export function Player (resource) {
  var player = new engine.Entity()
  player.addComponent('model')
  player.model.model = resource

  return player
}

/**
 *  Asteroid entity
 *  @params resource
 */
export function Asteroid (resource) {
  var asteroid = new engine.Entity()
  asteroid.addComponent('model')
  asteroid.model.model = resource

  return asteroid
}


src/renderer.js
import engine from 'engine'
import game from './game'

// define entities
import * as entity from './models/entities'
var camera, light, player, asteroid,
  radius = 20,
  height = 5,
  angle = 0

/**
 *  Game entities and mechanics setup
 *    - attaches starting entities to game
 *    - attaches camera and light entities
 *    - sets camera in default position
 *    - sets lighting in default state
 */
export function create () {

  // create default entities
  camera = new entity.Camera(),
  light = new entity.Light()

	// attach entities to screen
  game.root.addChild(camera)
  game.root.addChild(light)

  game.assets.loadFromUrl(entity.assets['shwip'], 'model')
    .then(function (results) {
      player = new entity.Player(results.resource)
      game.root.addChild(player)
    })

  game.assets.loadFromUrl(entity.assets['asteroid_1'], 'model')
    .then(function (results) {
      asteroid = new entity.Asteroid(results.resource)
      game.root.addChild(asteroid)
      asteroid.setPosition(10, 0, 0)
    })

  // camera default position
  camera.setPosition(0, 0, 30)

  // lighting default setting
  // light.setEulerAngles(45, 0, 0)

}

/**
 *  Game rendering loop
 *    - infinitely pans camera
 *    - infinitely rotates asteroid
 */
export function update (deltaTime) {

  // camera.rotateLocal(0, 3 * deltaTime, 30 * deltaTime)

  if (player) {
    light.lookAt(player.getPosition())
    camera.lookAt(player.getPosition())
    player.rotate(10 * deltaTime, 20 * deltaTime, 30 * deltaTime)
  }

  if (asteroid)
    asteroid.rotate(30 * deltaTime, 40 * deltaTime, 50 * deltaTime)

  angle = angle > 360 ? angle - 360 : angle + 20 * deltaTime
  light.rotateLocal(90, 0, 0)
  light.setLocalPosition(
    radius * Math.sin(angle*pc.math.DEG_TO_RAD),
    height,
    radius * Math.cos(angle*pc.math.DEG_TO_RAD)
  )

}
